src/Microsoft.ML.TimeSeries/SrCnnEntireAnomalyDetector.cs

```diff
@@ -0,0 +1,846 @@
+        protected override void ProcessBatch(Batch currentBatch)
+        {
+            currentBatch.Process();
+            currentBatch.Reset();
+        }
+
+        internal sealed class Batch
+        {
+            private List<double> _previousBatch;
+            private List<double> _batch;
+            private readonly int _outputLength;
+            private SrCnnEntireModeler _modeler;
+            private int _batchSize;
+            private double[][] _results;
+
+            public Batch(int batchSize, int outputLength, SrCnnEntireModeler modeler)
+            {
+                _batchSize = batchSize;
+                _outputLength = outputLength;
+                if (batchSize == -1)
+                {
+                    _previousBatch = new List<double>();
+                    _batch = new List<double>();
+                }
+                else
+                {
+                    _previousBatch = new List<double>(batchSize);
+                    _batch = new List<double>(batchSize);
+                }
+                _modeler = modeler;
+            }
+
+            public void AddValue(double value)
+            {
+                _batch.Add(value);
+            }
+
+            public int Count => _batch.Count;
+
+            public void Process()
+            {
+                _batchSize = _batch.Count;
+                if (_batch.Count < MinBatchSize)
+                {
+                    if (_previousBatch.Count + _batch.Count < MinBatchSize)
+                        return;
+                    var bLen = _previousBatch.Count - _batch.Count;
+                    _previousBatch = _previousBatch.GetRange(_batch.Count, bLen);
+                    _previousBatch.AddRange(_batch);
+                    _modeler.Train(_previousBatch.ToArray(), ref _results);
+                    _results = _results.Skip(bLen).ToArray();
+                }
+                else
+                {
+                    _modeler.Train(_batch.ToArray(), ref _results);
+                }
+            }
+
+            public void Reset()
+            {
+                var tempBatch = _previousBatch;
+                _previousBatch = _batch;
+                _batch = tempBatch;
+                _batch.Clear();
+            }
+
+            public ValueGetter<VBuffer<double>> CreateGetter(DataViewRowCursor input, string inputCol)
+            {
+                ValueGetter<double> srcGetter = input.GetGetter<double>(input.Schema[inputCol]);
+                ValueGetter<VBuffer<double>> getter =
+                    (ref VBuffer<double> dst) =>
+                    {
+                        double src = default;
+                        srcGetter(ref src);
+                        var result = VBufferEditor.Create(ref dst, _outputLength);
+                        for (int i = 0; i < _outputLength; ++i)
+                        {
+                            result.Values[i] = _results[input.Position % _batchSize][i];
+                        }
```

Comment:
Shorter alternative:
```
_results[...].CopyTo(result.Values)
```

Reply:
Fixed as you suggested.

---
In reply to: [427864940](https://github.com/dotnet/machinelearning/pull/5135#discussion_r427864940) [](ancestors = 427864940)